feat(timescaledb): create a TimescaleDB documentation (#1520)
Scalingo/documentation: _posts/databases/postgresql/timescaledb/2000-01-01-downsampling.md

---
title: Downsampling

> nav: Downsampling

Reviewer: You don't need a nav if it's the same as the title:
```suggestion
```

modified_at: 2022-01-21 00:00:00
tags: timescale databases postgresql extensions downsampling
---

> Downsampling data allows to reduce the amount of data by simplifying the old values.

Reviewer: Simplification suggestion:

```suggestion
The downsampling reduces the amount of data by simplifying the old values.
```

I don't like the word "simplifying". I don't understand what it means

## Example

> Bellow an example of a logic of downsampling. It is based on TimescaleDB documentation.

Reviewer: If it's based on the official doc, please make `TimescaleDB documentation` a link to the said page

```suggestion
Below an example of a logic of downsampling. It is based on TimescaleDB documentation.
```

The data is stored in a table `conditions`, composed by `temperature` and `humidity`

> value. This is a use case where a table is composed of raw datas.

Reviewer: data is still uncountable. Do you have a dictionary installed in the text editor you use to write doc? It should highlight that `datas` does not exist

```suggestion
value. This is a use case where a table is composed of raw data.
```

> Now we want to decrease the database size. We don't need to have data by the

Reviewer: nitpick: I'm not sure about the purpose of the word `Now` here. From my point of view it does not add any information.

```suggestion
We want to decrease the database size. We don't need to have data by the
```

minute for data older than one week, the average of values by hour is enough.

> So we can use a downsampling logic which will reduce the datas older than one week,
> and only store the average by hour.

Reviewer: We rarely use the future tense in documentation pages:
```suggestion
So we can use a downsampling logic which reduces the data older than one week,
and only stores the average by hour.
```

### Datas

> For the following table:

Reviewer: Not sure about this one

```suggestion
For the following data:
```

```
$ select * from conditions;
             time              | location |    temperature     |      humidity
-------------------------------+----------+--------------------+--------------------
 2022-01-21 09:50:05.049911+00 | office   |                 75 |                 52
 2022-01-21 09:50:05.058412+00 | office   |                 70 |                 57
 2022-01-21 09:50:05.059489+00 | office   |                 76 |                 52
 2022-01-20 09:50:05.06032+00  | office   |                 79 |                 54
```

### Downsample Procedure

> The procedure bellow contains the logic to downsample the table `conditions`

Reviewer: I didn't know but the word `bellow` exists and means "mugir" or "beugler" xD

```suggestion
The procedure below contains the logic to downsample the table `conditions`
```

      INNER JOIN pg_class pgc ON pgc.oid = show.oid
      INNER JOIN pg_namespace pgns ON pgc.relnamespace = pgns.oid
      INNER JOIN timescaledb_information.chunks chunk ON chunk.chunk_name = pgc.relname
        AND chunk.chunk_schema = pgns.nspname
  LOOP
    RAISE NOTICE 'Processing chunk: %', chunk::text;

    -- build name for temp table
    SELECT '_tmp' || relname
    FROM pg_class
    WHERE oid = chunk INTO STRICT tmp_name;

    -- copy downsampled chunk data into temp table
    EXECUTE format($sql$ CREATE UNLOGGED TABLE %I AS
      -- you can configure here the time range to aggregate datas and how you do it. Here we used `avg` function on 1 hour
      SELECT time_bucket('1h', time), location, avg(temperature) as avg_temperature, avg(humidity) as avg_humidity FROM %s GROUP BY 1, 2;
    $sql$, tmp_name, chunk);

    -- clear original chunk
    EXECUTE format('TRUNCATE %s;', chunk);

    -- copy downsampled data back into chunk
    EXECUTE format('INSERT INTO %s(time, location, temperature, humidity) SELECT * FROM %I;', chunk, tmp_name);

    -- drop temp table
    EXECUTE format('DROP TABLE %I;', tmp_name);

    COMMIT;
  END LOOP;
END
$$;
```

### Execute Procedure

> You can schedule the execution of the previous procedure by using

Reviewer: ```suggestion
You can schedule the execution of the previous procedure by using the
```

[Scalingo scheduler]({% post_url platform/app/task-scheduling/2000-01-01-scalingo-scheduler %}).

> To do so, add a `cron.json` file at the root of your application’s source code,

Reviewer: nitpick: not sure `To do so` adds anything here. And it really sounds like a French translation to my hears. I might be wrong:
```suggestion
Add a `cron.json` file at the root of your application’s source code,
```
